File: apps/web-next/src/lib/ml/fusion.ts
// ML Signal Fusion SDK v2.0.0
import { SLO_TARGETS } from "@/lib/constants/slo";
import { FEATURE_VERSION, MODEL_VERSION } from "./featureVersion";
import type { FeatureRow, ScoreRequest, ScoreResponse } from "./schema";

/**
 * Sanitize feature row - NaN/Infinity guard
 * Returns null if invalid
 */
export function sanitize(f: FeatureRow): FeatureRow | null {
  // Check required OHLCV
  const required = [f.o, f.h, f.l, f.c, f.v];
  if (!required.every(x => typeof x === "number" && Number.isFinite(x))) {
    return null;
  }
  
  // Check all numeric values are finite
  for (const [key, value] of Object.entries(f)) {
    if (typeof value === "number" && !Number.isFinite(value)) {
      return null;
    }
  }
  
  return f;
}

/**
 * Fuse signals - weighted ensemble
 * Returns score [-1, 1] and component parts
 */
export function fuseSignals(f: FeatureRow): { score: number; parts: Record<string, number> } {
  const parts: Record<string, number> = {};
  
  // RSI (mean-revert): oversold=bullish, overbought=bearish
  if (typeof f.rsi_14 === "number") {
    parts.rsi = (50 - f.rsi_14) / 50;  // normalize to [-1, 1]
  }
  
  // MACD histogram (momentum)
  if (typeof f.macd_hist === "number") {
    parts.macd = Math.tanh(f.macd_hist);  // squash to [-1, 1]
  }
  
  // EMA trend (crossover)
  if (typeof f.ema_20 === "number" && typeof f.ema_50 === "number" && f.c > 0) {
    parts.trend = Math.tanh((f.ema_20 - f.ema_50) / f.c);
  }
  
  // ATR volatility filter (optional)
  if (typeof f.atr_14 === "number" && f.c > 0) {
    const atrPct = f.atr_14 / f.c;
    parts.vol = atrPct > 0.05 ? 0 : 1;  // high vol → reduce confidence
  }
  
  // Weighted ensemble (weights sum to 1.0)
  const weights = { rsi: 0.3, macd: 0.4, trend: 0.3 };
  const score = weighted(parts, weights);
  
  return { score, parts };
}

/**
 * Weighted average helper
 */
function weighted(parts: Record<string, number>, w: Record<string, number>): number {
  let sum = 0;
  let totalWeight = 0;
  
  for (const key in parts) {
    const weight = w[key] ?? 0;
    sum += (parts[key] ?? 0) * weight;
    totalWeight += weight;
  }
  
  return totalWeight > 0 ? Math.max(-1, Math.min(1, sum / totalWeight)) : 0;
}

/**
 * Decision logic - convert score to discrete action
 */
export function decide(score: number, confidFloor = 0.55): { decision: -1 | 0 | 1; confid: number } {
  const absScore = Math.abs(score);
  const confid = Math.max(0, Math.min(1, absScore));
  
  // Below confidence threshold → flat (no action)
  if (absScore < confidFloor) {
    return { decision: 0, confid };
  }
  
  // Above threshold → long (1) or short (-1)
  const decision = score > 0 ? 1 : -1;
  return { decision: decision as -1 | 1, confid };
}

/**
 * Guardrails check - SLO compliance
 */
export function guardrailsCheck(ctx: {
  staleness_s?: number;
  p95_ms?: number;
  error_rate?: number;
  confid?: number;
}): { pass: boolean; reasons: string[] } {
  const reasons: string[] = [];
  
  // SLO checks
  if ((ctx.staleness_s ?? 0) > SLO_TARGETS.STALENESS_S) {
    reasons.push("staleness_breach");
  }
  if ((ctx.p95_ms ?? 0) > SLO_TARGETS.P95_MS) {
    reasons.push("p95_breach");
  }
  if ((ctx.error_rate ?? 0) > SLO_TARGETS.ERROR_RATE) {
    reasons.push("error_rate_breach");
  }
  
  // Confidence floor
  if ((ctx.confid ?? 0) < 0.55) {
    reasons.push("low_confidence");
  }
  
  return {
    pass: reasons.length === 0,
    reasons
  };
}

/**
 * Main scoring function
 */
export function score(req: ScoreRequest): ScoreResponse {
  // Sanitize input
  const f = sanitize(req.feature);
  if (!f) {
    throw new Error("invalid_feature: NaN or Infinity detected");
  }
  
  // Fuse signals
  const { score: rawScore, parts } = fuseSignals(f);
  
  // Decide action
  const { decision, confid } = decide(rawScore);
  
  // Guardrails check (can add SLO context here)
  const guard = guardrailsCheck({ confid });
  
  // Generate trace ID (will be replaced by OTEL in production)
  const traceId = typeof crypto !== 'undefined' && crypto.randomUUID 
    ? crypto.randomUUID() 
    : `ml-${Date.now()}-${Math.random().toString(16).slice(2, 10)}`;
  
  return {
    ts: f.ts,
    symbol: f.symbol,
    decision,
    score: rawScore,
    confid,
    featuresUsed: Object.keys(parts),
    modelId: req.modelId || MODEL_VERSION,
    featureVersion: FEATURE_VERSION,
    guardrails: guard,
    traceId
  };
}

/**
 * Batch scoring for backtesting
 */
export function scoreBatch(features: FeatureRow[], modelId: string): ScoreResponse[] {
  return features
    .map(f => {
      try {
        return score({ modelId, feature: f });
      } catch {
        return null;
      }
    })
    .filter((s): s is ScoreResponse => s !== null);
}

File: apps/web-next/src/lib/ml/fusion.py
"""
ML Signal Fusion - Python Reference Implementation
For training/backtesting (offline analysis)

Usage:
  from ml.fusion import sanitize, fuse, decide
  
  row = {...}  # FeatureRow dict
  clean = sanitize(row)
  if clean:
    score, parts = fuse(clean)
    decision, confid = decide(score)
"""

import numpy as np
from typing import Dict, Optional, Tuple, Literal

def sanitize(row: Dict) -> Optional[Dict]:
    """
    Validate feature row - reject NaN/Infinity
    
    Args:
        row: FeatureRow dict with OHLCV + indicators
        
    Returns:
        Cleaned row or None if invalid
    """
    # Check required OHLCV
    required = ["o", "h", "l", "c", "v"]
    for k in required:
        if k not in row or row[k] is None:
            return None
        if not np.isfinite(row[k]):
            return None
    
    # Check all numeric values are finite
    for k, v in row.items():
        if isinstance(v, (int, float)) and not np.isfinite(v):
            return None
    
    return row

def fuse(row: Dict) -> Tuple[float, Dict[str, float]]:
    """
    Fuse signals - weighted ensemble
    
    Args:
        row: Cleaned FeatureRow
        
    Returns:
        (score, parts) where score in [-1, 1]
    """
    parts = {}
    
    # RSI (mean-revert)
    if "rsi_14" in row and row["rsi_14"] is not None:
        parts["rsi"] = (50 - row["rsi_14"]) / 50
    
    # MACD histogram (momentum)
    if "macd_hist" in row and row["macd_hist"] is not None:
        parts["macd"] = np.tanh(row["macd_hist"])
    
    # EMA trend (crossover)
    if row.get("ema_20") and row.get("ema_50") and row.get("c", 0) > 0:
        parts["trend"] = np.tanh((row["ema_20"] - row["ema_50"]) / row["c"])
    
    # ATR volatility filter
    if row.get("atr_14") and row.get("c", 0) > 0:
        atr_pct = row["atr_14"] / row["c"]
        parts["vol"] = 0 if atr_pct > 0.05 else 1
    
    # Weighted average
    weights = {"rsi": 0.3, "macd": 0.4, "trend": 0.3}
    total_weight = sum(weights.get(k, 0) for k in parts.keys())
    
    if total_weight > 0:
        score = sum(parts.get(k, 0) * weights.get(k, 0) for k in parts.keys()) / total_weight
        score = max(-1, min(1, score))
    else:
        score = 0.0
    
    return score, parts

def decide(score: float, confid_floor: float = 0.55) -> Tuple[Literal[-1, 0, 1], float]:
    """
    Convert score to discrete decision
    
    Args:
        score: Continuous score in [-1, 1]
        confid_floor: Minimum confidence threshold
        
    Returns:
        (decision, confidence) where decision in {-1, 0, 1}
    """
    abs_score = abs(score)
    confid = max(0.0, min(1.0, abs_score))
    
    # Below confidence floor → flat
    if abs_score < confid_floor:
        return 0, confid
    
    # Above floor → directional
    decision = 1 if score > 0 else -1
    return decision, confid

# Example usage
if __name__ == "__main__":
    # Sample feature row
    row = {
        "ts": 1697241600000,
        "symbol": "BTCUSDT",
        "tf": "1h",
        "o": 27500, "h": 27650, "l": 27450, "c": 27600, "v": 1200,
        "rsi_14": 45.5,
        "macd_hist": 0.25,
        "ema_20": 27550,
        "ema_50": 27400,
        "atr_14": 150
    }
    
    clean = sanitize(row)
    if clean:
        score, parts = fuse(clean)
        decision, confid = decide(score)
        
        print(f"Score: {score:.3f}")
        print(f"Parts: {parts}")
        print(f"Decision: {decision} (confid: {confid:.3f})")
    else:
        print("Invalid row (NaN/Infinity)")

